### PR TITLE
initial prototype

### PR DESCRIPTION
**Pull request by angusmcleod**

### CODE REVIEW

File: assets/javascripts/discourse/initializers/preview-edits.js.es6 [hunk 1]
```diff
@@ -12,21 +12,37 @@ export default {
     if (!Discourse.SiteSettings.topic_list_previews_enabled) return;
 
     withPluginApi('0.8.12', (api) => {
-      api.modifyClass('component:topic-list', {
+
+      api.modifyClass('component:topic-list',  {
+        // we need 'div's for masonry
+        tagName: 'div',
```
[the remark this answers is not in the document]
Comment (merefield) [reply assets/javascripts/discourse/initializers/preview-edits.js.es6]: angus, if I remove this and function  switchTagBasedOnStyle' it breaks for me completely.  Had you tried that?

File: assets/javascripts/discourse/initializers/preview-edits.js.es6 [hunk 2]
```diff
@@ -12,21 +12,37 @@ export default {
...
         router: Ember.inject.service('-routing'),
         currentRoute: Ember.computed.alias('router.router.currentRouteName'),
-        classNameBindings: ['showThumbnail', 'showExcerpt', 'showActions', 'socialStyle'],
+        classNameBindings: ['showThumbnail', 'showExcerpt', 'showActions', 'socialStyle', 'tilesStyle'],
         suggestedList: Ember.computed.equal('parentView.parentView.parentView.elementId', 'suggested-topics'),
         discoveryList: Ember.computed.equal('parentView._debugContainerKey', 'component:discovery-topics-list'),
         listChanged: false,
 
+       switchTagBasedOnStyle: function () {
```
[the remark this answers is not in the document]
Comment (merefield) [reply assets/javascripts/discourse/initializers/preview-edits.js.es6]: thanks, cleaned up (removed)

File: assets/javascripts/discourse/initializers/preview-edits.js.es6 [hunk 3]
```diff
@@ -12,21 +12,37 @@ export default {
...
         setup() {
           const suggestedList = this.get('suggestedList');
+          this.switchTagBasedOnStyle();
           if (suggestedList) {
             const category = this.get('parentView.parentView.parentView.topic.category');
             this.set('category', category);
-          }
+          };
+          if (this.get('tilesStyle')){
+            Ember.run.scheduleOnce('afterRender', this, this.applyMasonry);
```
[the remark this answers is not in the document]
Comment (merefield) [reply assets/javascripts/discourse/initializers/preview-edits.js.es6]: as you suggested above, switchTagBasedOnStyle looks redundant as never called and the tag is apparently fine left as 'table'.  Removed the call completely.

File: assets/javascripts/discourse/initializers/preview-edits.js.es6 [hunk 4]
```diff
@@ -91,6 +107,16 @@ export default {
...
         },
 
+        @computed('listChanged')
+        tilesStyle() {
+          return this.settingEnabled('topic_list_tiles');
+        },
+
+        @computed('listChanged')
+        tilesOrSocial() {
+            return (this.get('tilesStyle') || this.get('socialStyle'));
```
[the remark this answers is not in the document]
Comment (merefield) [reply assets/javascripts/discourse/initializers/preview-edits.js.es6]: done

File: assets/javascripts/discourse/initializers/preview-edits.js.es6 [hunk 5]
```diff
@@ -12,21 +12,37 @@ export default {
...
         router: Ember.inject.service('-routing'),
         currentRoute: Ember.computed.alias('router.router.currentRouteName'),
-        classNameBindings: ['showThumbnail', 'showExcerpt', 'showActions', 'socialStyle'],
+        classNameBindings: ['showThumbnail', 'showExcerpt', 'showActions', 'socialStyle', 'tilesStyle'],
         suggestedList: Ember.computed.equal('parentView.parentView.parentView.elementId', 'suggested-topics'),
         discoveryList: Ember.computed.equal('parentView._debugContainerKey', 'component:discovery-topics-list'),
         listChanged: false,
 
+       switchTagBasedOnStyle: function () {
+            this.tagName = 'tr';
```
[the remark this answers is not in the document]
Comment (merefield) [reply assets/javascripts/discourse/initializers/preview-edits.js.es6]: haha, thanks, had already done this in new version as broke the layout of the standard view.  Function removed as per above in any case.

File: assets/javascripts/discourse/initializers/preview-edits.js.es6 [hunk 6]
```diff
@@ -113,33 +139,81 @@ export default {
...
         @computed('listChanged')
         thumbnailFirstXRows() {
           return Discourse.SiteSettings.topic_list_thumbnail_first_x_rows;
-        }
+        },
+
+        // don't forget to update masonry layout when required
+        masonryObserver: function() {
+             Ember.run.scheduleOnce('afterRender', this, this.applyMasonry);
+      	}.observes("topics.[]"),
```
[the remark this answers is not in the document]
Comment (merefield) [reply assets/javascripts/discourse/initializers/preview-edits.js.es6]: updated to new style, thanks!

File: assets/javascripts/discourse/initializers/preview-edits.js.es6 [hunk 7]
```diff
@@ -113,33 +139,81 @@ export default {
...
+            msnry = this.$('.grid').data('masonry')
+            $('.grid').imagesLoaded(function() {msnry.layout()});
+          };
+          }
       });
 
       api.modifyClass('component:topic-list-item', {
+        // needs 'div's for masonry
+        tagName: 'div',
+        classNames: ['grid-item'],
```
[the remark this answers is not in the document]
Comment (merefield) [reply assets/javascripts/discourse/initializers/preview-edits.js.es6]: got it!  removed

File: assets/javascripts/discourse/initializers/preview-edits.js.es6 [hunk 8]
```diff
@@ -113,33 +139,81 @@ export default {
...
+        tilesOrSocial: Ember.computed.alias('parentView.tilesOrSocial'),
         showThumbnail: Ember.computed.and('thumbnails', 'parentView.showThumbnail'),
         showExcerpt: Ember.computed.and('topic.excerpt', 'parentView.showExcerpt'),
         showActions: Ember.computed.alias('parentView.showActions'),
         showCategoryBadge: Ember.computed.alias('parentView.showCategoryBadge'),
         thumbnailFirstXRows: Ember.computed.alias('parentView.thumbnailFirstXRows'),
         category: Ember.computed.alias('parentView.category'),
 
+        switchTagBasedOnStyle: function() {
+          if (this.get('tilesStyle') == false) {
```
[the remark this answers is not in the document]
Comment (merefield) [reply assets/javascripts/discourse/initializers/preview-edits.js.es6]: great.  Simplified and moved into the _setupProperties function, getting rid of the additional method.

File: assets/javascripts/discourse/templates/list/topic-list-item.raw.hbs [hunk 9]
```diff
@@ -1,3 +1,54 @@
+{{#if tilesStyle}}
```
[the remark this answers is not in the document]
Comment (merefield) [reply assets/javascripts/discourse/templates/list/topic-list-item.raw.hbs]: integrated with main body.  I think this is a trade off ... there are a lot of bitty conditionals now, making it harder to visualise a single 'theme style', but we've saved a few lines.  will submit the updated version in any case. thanks

File: assets/javascripts/discourse/templates/components/topic-list.hbs [hunk 10]
```diff
@@ -0,0 +1,69 @@
+{{#unless skipHeader}}
```
[the remark this answers is not in the document]
Comment (merefield) [reply assets/javascripts/discourse/templates/components/topic-list.hbs]: angus, I understand your nervousness, but according to this:

https://stackoverflow.com/questions/918792/use-jquery-to-change-an-html-tag/20469901

"Once a tag is created in the DOM it's immutable" ... so not sure I can change them without major hack as shown on this page.  Do we want to go down that route?

replaceWith() just removes all the matched content (including all the internal stuff) and replaces it with what you provide ... deleting all the internal detail if you simply do this:

$('tbody').replaceWith("<div class='grid' >");

In additional there is more than one 'tbody' on the page potentially and there is no identifier to isolate this one?

ideas?

File: assets/stylesheets/previews_common.scss [hunk 11]
```diff
@@ -1,397 +1,528 @@
-.topic-thumbnail {
```
[the remark this answers is not in the document]
Comment (merefield) [reply assets/stylesheets/previews_common.scss]: this is not yet addressed in any submitted code, but FYI my styling on the tiles is internally based on CSS Grid which requires significantly different CSS so felt it right to break it out completely.

File: assets/javascripts/discourse/templates/components/topic-list.hbs (hunk 10 above)
[the remark this answers is not in the document]
Comment (angusmcleod) [reply assets/javascripts/discourse/templates/components/topic-list.hbs]: I just tested it out without changing the tag, and just adding a class, and it seems to be working. Try adding ``grid`` to the ``tbody`` and adding a ``display: block`` to ``grid`` style.

[the remark this answers is not in the document]
Comment (merefield) [reply assets/javascripts/discourse/templates/components/topic-list.hbs]: Thanks!  Done! This seems to work most of the time ... but has introduced an intermittent issue where the 'There are no more latest topics.' overlaps the first item ... let's keep an eye on that.

**Comment by angusmcleod** [comment]
Closing in favour of: https://github.com/angusmcleod/discourse-topic-previews/pull/33

**Pull request closed without merging**